cmd/nitro/init.go

```diff
@@ -121,6 +142,97 @@ func downloadInit(ctx context.Context, initConfig *conf.InitConfig) (string, err
 	}
 }
 
+// fetchChecksum performs a GET request to the specified URL using the provided context
+// and returns the checksum as a []byte
+func fetchChecksum(ctx context.Context, url string) ([]byte, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error making GET request: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, notFoundError
+	} else if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %v", resp.Status)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response body: %w", err)
+	}
+	checksumStr := strings.TrimSpace(string(body))
+	checksum, err := hex.DecodeString(checksumStr)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding checksum: %w", err)
+	}
+	if len(checksum) != sha256.Size {
+		return nil, fmt.Errorf("invalid checksum length")
+	}
+	return checksum, nil
+}
+
+func downloadInitInParts(ctx context.Context, initConfig *conf.InitConfig) (string, error) {
+	log.Info("File not found; trying to download database in parts")
+	fileInfo, err := os.Stat(initConfig.DownloadPath)
+	if err != nil || !fileInfo.IsDir() {
+		return "", fmt.Errorf("download path must be a directory: %v", initConfig.DownloadPath)
+	}
+	part := 0
+	parts := []string{}
+	defer func() {
+		// remove all temporary files.
+		for _, part := range parts {
+			err := os.Remove(part)
+			if err != nil {
+				log.Warn("Failed to remove temporary file", "file", part)
+			}
+		}
+	}()
+	for {
+		url := fmt.Sprintf("%s.part%d", initConfig.Url, part)
+		log.Info("Downloading database part", "url", url)
+		partFile, err := downloadFile(ctx, initConfig, url)
+		if errors.Is(err, notFoundError) {
+			log.Info("Part not found; concatenating archive into single file")
+			break
+		} else if err != nil {
+			return "", err
+		}
+		parts = append(parts, partFile)
+		part++
+	}
+	return joinArchive(parts)
+}
+
+// joinArchive joins the archive parts into a single file and return its path.
+func joinArchive(parts []string) (string, error) {
+	if len(parts) == 0 {
+		return "", fmt.Errorf("no database parts found")
+	}
+	archivePath := strings.TrimSuffix(parts[0], ".part0")
+	archive, err := os.Create(archivePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to create archive: %w", err)
+	}
+	defer archive.Close()
+	for _, part := range parts {
+		partFile, err := os.Open(part)
+		if err != nil {
+			return "", fmt.Errorf("failed to open part file %s: %w", part, err)
+		}
+		defer partFile.Close()
+		_, err = io.Copy(archive, partFile)
```

Comment:
This could also take a while, could this make an info log for each part appended?